blocks/form/components/repeat/repeat.js

```diff
@@ -38,45 +59,143 @@ function createButton(label, icon) {
+  const instances = wrapper.querySelectorAll('[data-repeatable="true"]');
+
+  instances.forEach((instance) => {
+    if (checkExisting) {
+      const existingRemoveButton = instance.querySelector('.item-remove');
+      if (existingRemoveButton) {
+        return;
+      }
+    }
+
+    insertRemoveButton(instance, wrapper, form, isDocBased);
+  });
+}
+
+function setupModelSubscription(wrapper, form, formId) {
+  const containerElement = wrapper.closest('fieldset[data-id]');
+
+  subscribe(containerElement, formId, (fieldDiv, fieldModel) => {
+    wrapper.fieldModel = fieldModel;
+    fieldModel.subscribe((e) => {
+      const { payload } = e;
+      payload?.changes?.forEach((change) => {
+        if (change?.propertyName === 'items') {
+          // eslint-disable-next-line max-len
+          // Reason for requestAnimationFrame: Model changes fire immediately but DOM updates are async.
+          // We need to wait for the browser's next paint cycle to ensure the new/removed fieldsets
+          // are in the DOM before adding/updating buttons.
+          requestAnimationFrame(() => {
+            wrapper.querySelectorAll('[data-repeatable="true"]').forEach((instance, index) => {
+              updateRadioButtonNames(instance, index);
+            });
+            addRemoveButtons(wrapper, form, false, true);
+            updateButtonVisibility(wrapper);
+          });
+        }
+      });
+    }, 'change');
+  });
+}
+
+// Doc-based manual DOM manipulation (exception case)
+function addDocBasedInstance(wrapper, form) {
   const fieldset = wrapper['#repeat-template'];
-  const max = wrapper.getAttribute('data-max');
-  const min = wrapper.getAttribute('data-min');
   const childCount = wrapper.children.length - 1;
   const newFieldset = fieldset.cloneNode(true);
   newFieldset.setAttribute('data-index', childCount);
   update(newFieldset, childCount, wrapper['#repeat-template-label']);
-  if (childCount >= +min) {
-    insertRemoveButton(newFieldset, wrapper, form);
-  }
-  if (+max !== -1 && +max <= childCount + 1) {
-    e.currentTarget.setAttribute('data-visible', 'false');
-  }
+
+  const actions = wrapper.querySelector('.repeat-actions');
   actions.insertAdjacentElement('beforebegin', newFieldset);
+
+  // Add remove button to the new instance
+  insertRemoveButton(newFieldset, wrapper, form, true);
+
+  // Add remove buttons to all existing instances that don't have them
+  // (this handles the case where we started with min instances and no buttons)
+  addRemoveButtons(wrapper, form, true, true);
+
+  updateButtonVisibility(wrapper);
+
   const event = new CustomEvent('item:add', {
     detail: { item: { name: newFieldset.name, id: newFieldset.id } },
     bubbles: false,
   });
   form.dispatchEvent(event);
-};
+}
+
+function addInstance(wrapper, form, isDocBased = false) {
+  if (isDocBased) {
+    addDocBasedInstance(wrapper, form);
+  } else {
+    addAFBasedInstance(wrapper);
+  }
```

Comment:
I would suggest the following
create the following code structure
repeat.js
repeat-doc.js
repeat-af.js
repeat-common.js

This could avoid two many if checks

Reply:
I've used a simpler approach within one file, sort of a strategy pattern, please review.